lib/puppet/type/firewalld_policy.rb

```diff
@@ -0,0 +1,308 @@
+  #
+  Puppet::Type.type(:firewalld_rich_rule)
+  Puppet::Type.type(:firewalld_service)
+  Puppet::Type.type(:firewalld_port)
+
+  desc <<-DOC
+    @summary
+      Creates and manages firewalld policies.
+
+    Creates and manages firewalld policies.
+
+    Note that setting `ensure => 'absent'` to the built in firewalld
+    policies will not work, and will generate an error. This is a
+    limitation of firewalld itself, not the module.
+
+    @example Create a policy called `anytorestricted`
+      firewalld_policy { 'anytorestricted':
+        ensure           => present,
+        target           => '%%REJECT%%',
+        ingress_zones    => ['ANY'],
+        egress_zones     => ['restricted'],
+        purge_rich_rules => true,
+        purge_services   => true,
+        purge_ports      => true,
+        icmp_blocks      => 'router-advertisement'
+      }
+  DOC
+
+  ensurable do
+    defaultvalues
+    defaultto :present
+  end
+
+  # When set to 1 these variables cause the purge_* options to
+  # indicate to Puppet that we are in a changed state
+  #
+  attr_reader :rich_rules_purgable
+  attr_reader :services_purgable
+  attr_reader :ports_purgable
+
+  def generate
+    return [] unless Puppet::Provider::Firewalld.available?
+    purge_rich_rules if self[:purge_rich_rules] == :true
+    purge_services if self[:purge_services] == :true
+    purge_ports if self[:purge_ports] == :true
+    []
+  end
+
+  newparam(:name) do
+    desc 'Name of the rule resource in Puppet'
+    isnamevar
+  end
+
+  newparam(:policy) do
+    desc 'Name of the policy'
+  end
+
+  newparam(:description) do
+    desc 'Description of the policy to add'
+  end
+
+  newparam(:short) do
+    desc 'Short description of the policy to add'
+  end
+
+  newproperty(:target) do
+    desc 'Specify the target for the policy'
+  end
+
+  newproperty(:ingress_zones, array_matching: :all) do
+    desc 'Specify the ingress zones for the policy'
+
+    validate do |value|
+      return if value == :unset
+
+      if value.length.zero?
+        raise Puppet::Error, 'parameter ingress_zones must contain at least one zone'
+      end
+
+      return if value.length == 1
+
+      if value.include?('HOST') || value.include?('ANY')
+        raise Puppet::Error, 'parameter ingress_zones must contain a single symbolic zone or one or more regular zones'
+      end
+    end
+
+    def insync?(is)
+      case should
+      when String then should.lines.sort == is.sort
+      when Array then should.sort == is.sort
+      else raise Puppet::Error, 'parameter ingress_zones must be a string or array of strings!'
+      end
+    end
+  end
+
+  newproperty(:egress_zones, array_matching: :all) do
+    desc 'Specify the egress zones for the policy'
+
+    validate do |value|
+      return if value == :unset
+
+      if value.length.zero?
+        raise Puppet::Error, 'egress egress_zones must contain at least one zone'
+      end
+
+      return if value.length == 1
```

Comment:
`value` can be a string here if passed `['ANY']`, and the length will be the string length instead of the array length. This check does not work as expected.

Reply:
Fixed in recently pushed commit but not tested beyond the spec tests and not rebased.